working/2509-private-named-parameters/feature-specification.md

```diff
@@ -0,0 +1,382 @@
+*   If *p* has no corresponding public name *n*, then compile-time error. *You
+    can't use a private name for a named parameter unless there is a valid
+    public name that could be used at the callsite.*
+
+*   If any other parameter in C has declared name *p* or *n*, then
+    compile-time error. *If removing the `_` leads to a collision with
+    another parameter, then there is a conflict.*
+
+*   Otherwise, the name of the parameter in C is *n*. *If the parameter is
+    named, this then avoids the compile-time error that would otherwise be
+    reported for a private named parameter.*
+
+*   The local variable in the initializer list scope of C is *p*. *In the
+    initializer list, the private name is used. Inside the body of the
+    constructor, uses of *p* refer to the field, not the parameter.*
+
+*   If the parameter is an initializing formal, then it initializes a
+    corresponding field with name *p*.
+
+*   Else the field parameter induces an instance field with name *p*.
+
+*For example:*
+
+```dart
+class Id {
+  late final int _region = 0;
+
+  this({this._region, final int _value}) : assert(_region > 0 && _value > 0);
+
+  @override
+  String toString() => 'Id($_region, $_value)';
+}
+
+main() {
+  print(Id(region: 1, value: 2)); // Prints "Id(1, 2)".
+}
+```
+
+*Note that the proposal only applies to initializing formals and field
+parameters. A named parameter can only have a private name in a context where it
+is _useful_ to do so because it corresponds to a private instance field. For all
+other named parameters it is still a compile-time error to have a private name.*
+
+## Runtime semantics
+
+There are no runtime semantics for this feature. It's purely a compile-time
+renaming.
+
+## Compatibility
+
+This proposal takes code that it is currently a compile-time error (a private
+named parameter) and makes it valid in some circumstances (when the named
+parameter is an initializing formal or field parameter). Since it simply expands
+the set of valid programs, it is backwards compatible. Even so, it should be
+language versioned so that users don't inadvertently use this feature while
+their program allows being run on older pre-feature SDKs.
+
+## Tooling
+
+The best language features are designed holistically with the entire user
+experience in mind, including tooling and diagnostics. This section is *not
+normative*, but is merely suggestions and ideas for the implementation teams.
+They may wish to implement all, some, or none of this, and will likely have
+further ideas for additional warnings, lints, and quick fixes.
+
+### API documentation generation
+
+Docs generated using [`dart doc`][dartdoc] or other documentation generators
+should document any private named parameters using their public name. The fact
+that the parameter initializes or declares a private field is an implementation
+detail of the class. What a user of the class cares about is the corresponding
+public name for the constructor parameter.
```

Comment:
More importantly: *Authors* should document the public name. You write `// The [foo] is ...` to document `this._foo`.

The generated documentation should follow suit.
The tools may support `[_foo]` as well, and display it as <u>`foo`</u>, but the important thing is what authors should write in the doc source.